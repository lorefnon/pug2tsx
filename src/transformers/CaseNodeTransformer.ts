import * as Pug from "../pug";
import * as t from "@babel/types";
import { Transformer } from "./Transformer";
import { compact, flatten } from "lodash";
import { parseExpression } from "@babel/parser";
import { ErrorCode } from "../CompilationError";
import { FunctionBodyTransformer } from "./FunctionBodyTransformer";
import { PugNodeTransformer } from "./PugNodeTransformer";

export class CaseNodeTransformer extends Transformer<Pug.CaseNode, t.Node> {
    transform(): void {
        try {
            this.output = t.callExpression(
                t.arrowFunctionExpression(
                    [],
                    t.blockStatement([
                        t.switchStatement(
                            parseExpression(this.input.expr),
                            compact(
                                this.input.block.nodes.map(n => {
                                    if (!Pug.isType<Pug.WhenNode>(n, Pug.Type.When)) {
                                        this.pushError({
                                            reasons: ["Expected only when node in case statement body"],
                                            isFatal: true,
                                            code: ErrorCode.UnsupportedSyntaxError,
                                        });
                                        return;
                                    }
                                    const transformed = flatten(
                                        compact(n.block.nodes.map(n => this.delegateTo(PugNodeTransformer, n))),
                                    );
                                    const body = this.delegateTo(FunctionBodyTransformer, transformed);
                                    if (!body) return;
                                    if (n.expr === "default") return t.switchCase(null, body);
                                    else return t.switchCase(parseExpression(n.expr), body);
                                }),
                            ),
                        ),
                    ]),
                ),
                [],
            );
            this.commentLineNumber(this.output, this.input)
        } catch (e) {
            this.pushError({
                code: ErrorCode.UnsupportedSyntaxError,
                originalError: e,
                isFatal: true,
            });
        }
    }
}
