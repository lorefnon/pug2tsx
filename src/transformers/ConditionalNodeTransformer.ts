import { parseExpression } from "@babel/parser";
import * as t from "@babel/types";

import { ErrorCode } from "../CompilationError";
import * as Pug from "../pug";
import { Transformer } from "./Transformer";
import { NodeChildTransformer } from "./NodeChildTransformer";

export class ConditionalNodeTransformer extends Transformer<Pug.ConditionalNode, t.Node> {
    transform(): void {
        try {
            this.output = t.conditionalExpression(
                parseExpression(this.input.test),
                this.delegateTo(NodeChildTransformer, this.input.consequent) || t.identifier("undefined"),
                (this.input.alternate && this.delegateTo(NodeChildTransformer, this.input.alternate)) ||
                    t.identifier("undefined"),
            );
            this.commentLineNumber(this.output, this.input);
        } catch (e) {
            this.pushError({
                code: ErrorCode.UnsupportedSyntaxError,
                originalError: e,
                isFatal: true,
            });
        }
    }
}
