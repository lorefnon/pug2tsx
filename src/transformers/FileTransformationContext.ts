import * as t from "@babel/types";
import { some } from "lodash";
import { CompilationError } from "../CompilationError";

export class FileTransformationContext<ErrT = CompilationError> {
    public topLevelStatements: t.Statement[] = [];
    public errors: ErrT[] = [];

    constructor(
        public filePath?: string,
        public skipLineAnnotations?: boolean
    ) {}

    shouldGenerate() {
        return !some(this.errors, { isFatal: true });
    }
}
