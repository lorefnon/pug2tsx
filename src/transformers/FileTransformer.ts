import generateJS from "@babel/generator";
import * as t from "@babel/types";
import prettier from "prettier";
import { parse } from "../parser";
import { lex } from "../lexer";

import * as Pug from "../pug";
import { PugASTTransformer } from "./PugASTTransformer";
import { Transformer } from "./Transformer";
import { ErrorCode } from "../CompilationError";

export const parsePug = (input: string, context: { filePath?: string } = {}) => {
    const tokens = lex(input, { filename: context.filePath });
    return parse(tokens, { filename: context.filePath, src: input });
};

export class FileTransfomer extends Transformer<string, string> {
    transform() {
        const ast = this.parseAST();
        if (!ast) return;
        const transformer = new PugASTTransformer(ast, this.context);
        transformer.transform();
        const statements = transformer.output;
        if (!statements) return;
        this.output = this.serialize(statements);
    }
    parseAST() {
        let ast;
        try {
            ast = parsePug(this.input, this.context);
        } catch (e) {
            this.pushError({
                code: ErrorCode.PugParseError,
                originalError: e,
                reasons: ["Failed to parse Pug Template"],
                isFatal: true,
            });
            return;
        }
        if (!Pug.isType<Pug.BlockNode>(ast, Pug.Type.Block)) {
            this.pushError({
                code: ErrorCode.UnsupportedSyntaxError,
                reasons: ["Expected top level to be a block"],
                column: 1,
                line: 1,
                isFatal: true,
            });
            return;
        }
        return ast;
    }
    serialize(nodes: t.Node[]) {
        let result: string;
        try {
            result = nodes.map(n => generateJS(n).code).join("\n\n");
        } catch (e) {
            this.pushError({
                code: ErrorCode.GenerationError,
                reasons: ["Failed to generate javascript from transformed AST"],
                maybeBug: true,
                originalError: e,
                isFatal: true,
            });
            return;
        }
        try {
            result = prettier.format(result, {
                parser: "babel",
            });
        } catch (e) {
            console.log('Generated js code: ')
            console.log(result);
            this.pushError({
                code: ErrorCode.GenerationError,
                reasons: ["Failed to prettify generated javascript"],
                isFatal: false,
            });
        }
        return result;
    }
}
