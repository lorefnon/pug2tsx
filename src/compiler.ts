import { FileTransformationContext } from "./transformers/FileTransformationContext";
import { FileTransfomer } from "./transformers/FileTransformer";

export interface CompilationOptions {
    filePath?: string;
    skipLineAnnotations?: boolean
}

export const transpile = (inputContent: string, options: CompilationOptions = {}) => {
    const context = new FileTransformationContext(
        options.filePath,
        options.skipLineAnnotations
    );
    const transformer = new FileTransfomer(inputContent, context);
    transformer.transform();
    return {
        result: transformer.output,
        errors: context.errors,
    };
};
