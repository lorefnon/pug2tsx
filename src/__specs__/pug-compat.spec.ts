import {transpile} from "../compiler";
import dedent from "dedent";
import {isEmpty} from "lodash";
import {parsePug} from "../transformers/FileTransformer";

describe("Attributes", () => {
    test("Attribute expressions", () => {
        const template = dedent`
        - const Foo = () =>
            a(class='button' href='google.com') Google
            |
            |
            a(class='button', href='google.com') Google
            - var authenticated = true
            div(class=(authenticated ? 'authed' : 'anon'))
        `;
        expect(parsePug(template)).toMatchSnapshot();
        const transpiledR = transpile(template, {});
        expect(transpiledR).toMatchSnapshot();
        expect(isEmpty(transpiledR.errors)).toBe(true);
    });

    test("Multiline attributes", () => {
        const template = dedent`
        - const Foo = () =>
            input(
                type='checkbox'
                name='agreement'
                checked
            )
        `;
        expect(parsePug(template)).toMatchSnapshot();
        const transpiledR = transpile(template, {});
        expect(transpiledR).toMatchSnapshot();
        expect(isEmpty(transpiledR.errors)).toBe(true);
    });

    test("Quoted attributes", () => {
        const template = dedent`
        - const Foo = () =>
            div(class='div-class', (click)='play()')
            div(class='div-class' '(click)'='play()')`;
        expect(parsePug(template)).toMatchSnapshot();
        const transpiledR = transpile(template, {});
        expect(transpiledR).toMatchSnapshot();
        expect(isEmpty(transpiledR.errors)).toBe(true);
    });

    // TODO: Unescaped attributes
    // TODO: Boolean attributes

    test("Style attributes", () => {
        const template = dedent`
        - const Foo = () =>
            a(style={color: 'red', background: 'green'})`;
        expect(parsePug(template)).toMatchSnapshot();
        const transpiledR = transpile(template, {});
        expect(transpiledR).toMatchSnapshot();
        expect(isEmpty(transpiledR.errors)).toBe(true);
    });
    // TODO Autoconvert style string attributes
    // TODO Auto convert from dash to camelcase

    // TODO various cases of merging classes
    // TOOD Attribute blocks
});

describe("Case", () => {
    test("Default usage", () => {
        const template = dedent`
        - const Foo = () => 
            - var friends = 10
            case friends
                when 0
                    p you have no friends
                when 1
                    p you have a friend
                default
                    p you have #{friends} friends`;
        expect(parsePug(template)).toMatchSnapshot();
        const transpiledR = transpile(template, {});
        expect(transpiledR).toMatchSnapshot();
        expect(isEmpty(transpiledR.errors)).toBe(true);
    });
});

describe("Conditional", () => {
    test("Default usage", () => {
        const template = dedent`
        - const Foo = () =>
            - var user = { description: 'foo bar baz' }
            - var authorised = false
            #user
                if user.description
                    h2.green Description
                    p.description= user.description
                else if authorised
                    h2.blue Description
                    p.description.
                        User has no descriptio
                        why not add one...
                else
                    h2.red Description
                    p.description User has no description`;
        expect(parsePug(template)).toMatchSnapshot();
        const transpiledR = transpile(template, {});
        expect(transpiledR).toMatchSnapshot();
        expect(isEmpty(transpiledR.errors)).toBe(true);
    });
});

describe("Each", () => {
    test("Iteration with member", () => {
        const template = dedent`
        - const Foo = () =>
            ul
                each val in [1, 2, 3, 4, 5]
                    li= val`;
        expect(parsePug(template)).toMatchSnapshot();
        const transpiledR = transpile(template, {});
        expect(transpiledR).toMatchSnapshot();
        expect(isEmpty(transpiledR.errors)).toBe(true);
    });
    test("Iteration with member and index", () => {
        const template = dedent`
        - const Foo = () =>
            ul
                each val, index in ['zero', 'one', 'two']
                    li= index + ': ' + val`;
        expect(parsePug(template)).toMatchSnapshot();
        const transpiledR = transpile(template, {});
        expect(transpiledR).toMatchSnapshot();
        expect(isEmpty(transpiledR.errors)).toBe(true);
    });
    test("Iteration over object", () => {
        const template = dedent`
        - const Foo = () =>
            ul
                each val, index in {1:'one',2:'two',3:'three'}
                    li= index + ': ' + val`;
        expect(parsePug(template)).toMatchSnapshot();
        const transpiledR = transpile(template, {});
        expect(transpiledR).toMatchSnapshot();
        expect(isEmpty(transpiledR.errors)).toBe(true);
    });
});
