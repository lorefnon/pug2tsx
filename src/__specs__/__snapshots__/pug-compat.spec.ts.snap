// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Attributes Attribute expressions 1`] = `
{
  "line": 0,
  "nodes": [
    {
      "block": {
        "line": 2,
        "nodes": [
          {
            "attributeBlocks": [],
            "attrs": [
              {
                "column": 7,
                "line": 2,
                "mustEscape": true,
                "name": "class",
                "val": "'button'",
              },
              {
                "column": 22,
                "line": 2,
                "mustEscape": true,
                "name": "href",
                "val": "'google.com'",
              },
            ],
            "block": {
              "line": 2,
              "nodes": [
                {
                  "column": 41,
                  "line": 2,
                  "type": "Text",
                  "val": "Google",
                },
              ],
              "type": "Block",
            },
            "column": 5,
            "isInline": true,
            "line": 2,
            "name": "a",
            "selfClosing": false,
            "type": "Tag",
          },
          {
            "column": 6,
            "line": 3,
            "type": "Text",
            "val": "",
          },
          {
            "column": 1,
            "line": 4,
            "type": "Text",
            "val": "
",
          },
          {
            "column": 6,
            "line": 4,
            "type": "Text",
            "val": "",
          },
          {
            "attributeBlocks": [],
            "attrs": [
              {
                "column": 7,
                "line": 5,
                "mustEscape": true,
                "name": "class",
                "val": "'button'",
              },
              {
                "column": 23,
                "line": 5,
                "mustEscape": true,
                "name": "href",
                "val": "'google.com'",
              },
            ],
            "block": {
              "line": 5,
              "nodes": [
                {
                  "column": 42,
                  "line": 5,
                  "type": "Text",
                  "val": "Google",
                },
              ],
              "type": "Block",
            },
            "column": 5,
            "isInline": true,
            "line": 5,
            "name": "a",
            "selfClosing": false,
            "type": "Tag",
          },
          {
            "buffer": false,
            "column": 5,
            "isInline": false,
            "line": 6,
            "mustEscape": false,
            "type": "Code",
            "val": "var authenticated = true",
          },
          {
            "attributeBlocks": [],
            "attrs": [
              {
                "column": 9,
                "line": 7,
                "mustEscape": true,
                "name": "class",
                "val": "(authenticated ? 'authed' : 'anon')",
              },
            ],
            "block": {
              "line": 7,
              "nodes": [],
              "type": "Block",
            },
            "column": 5,
            "isInline": false,
            "line": 7,
            "name": "div",
            "selfClosing": false,
            "type": "Tag",
          },
        ],
        "type": "Block",
      },
      "buffer": false,
      "column": 1,
      "isInline": false,
      "line": 1,
      "mustEscape": false,
      "type": "Code",
      "val": "const Foo = () =>",
    },
  ],
  "type": "Block",
}
`;

exports[`Attributes Attribute expressions 2`] = `
{
  "errors": [],
  "result": "/*@pug:L1*/
const Foo = () => {
  return (
    <>
      {" "}
      /*@pug:L2*/
      <a href={"google.com"} className={"button"}>
        Google
      </a>
      /*@pug:L5*/
      <a href={"google.com"} className={"button"}>
        Google
      </a>
      {(() => {
        /*@pug:L6*/
        var authenticated = true;
        return (
          /*@pug:L7*/ <div className={authenticated ? "authed" : "anon"}></div>
        );
      })()}
    </>
  );
};
",
}
`;

exports[`Attributes Multiline attributes 1`] = `
{
  "line": 0,
  "nodes": [
    {
      "block": {
        "line": 2,
        "nodes": [
          {
            "attributeBlocks": [],
            "attrs": [
              {
                "column": 9,
                "line": 3,
                "mustEscape": true,
                "name": "type",
                "val": "'checkbox'",
              },
              {
                "column": 9,
                "line": 4,
                "mustEscape": true,
                "name": "name",
                "val": "'agreement'",
              },
              {
                "column": 9,
                "line": 5,
                "mustEscape": true,
                "name": "checked",
                "val": true,
              },
            ],
            "block": {
              "line": 2,
              "nodes": [],
              "type": "Block",
            },
            "column": 5,
            "isInline": false,
            "line": 2,
            "name": "input",
            "selfClosing": false,
            "type": "Tag",
          },
        ],
        "type": "Block",
      },
      "buffer": false,
      "column": 1,
      "isInline": false,
      "line": 1,
      "mustEscape": false,
      "type": "Code",
      "val": "const Foo = () =>",
    },
  ],
  "type": "Block",
}
`;

exports[`Attributes Multiline attributes 2`] = `
{
  "errors": [],
  "result": "/*@pug:L1*/
const Foo = () => (
  /*@pug:L2*/ <input
    type={"checkbox"}
    name={"agreement"}
    checked={true}
  ></input>
);
",
}
`;

exports[`Attributes Quoted attributes 1`] = `
{
  "line": 0,
  "nodes": [
    {
      "block": {
        "line": 2,
        "nodes": [
          {
            "attributeBlocks": [],
            "attrs": [
              {
                "column": 9,
                "line": 2,
                "mustEscape": true,
                "name": "class",
                "val": "'div-class'",
              },
              {
                "column": 28,
                "line": 2,
                "mustEscape": true,
                "name": "(click)",
                "val": "'play()'",
              },
            ],
            "block": {
              "line": 2,
              "nodes": [],
              "type": "Block",
            },
            "column": 5,
            "isInline": false,
            "line": 2,
            "name": "div",
            "selfClosing": false,
            "type": "Tag",
          },
          {
            "attributeBlocks": [],
            "attrs": [
              {
                "column": 9,
                "line": 3,
                "mustEscape": true,
                "name": "class",
                "val": "'div-class'",
              },
              {
                "column": 27,
                "line": 3,
                "mustEscape": true,
                "name": "(click)",
                "val": "'play()'",
              },
            ],
            "block": {
              "line": 3,
              "nodes": [],
              "type": "Block",
            },
            "column": 5,
            "isInline": false,
            "line": 3,
            "name": "div",
            "selfClosing": false,
            "type": "Tag",
          },
        ],
        "type": "Block",
      },
      "buffer": false,
      "column": 1,
      "isInline": false,
      "line": 1,
      "mustEscape": false,
      "type": "Code",
      "val": "const Foo = () =>",
    },
  ],
  "type": "Block",
}
`;

exports[`Attributes Quoted attributes 2`] = `
{
  "errors": [],
  "result": "/*@pug:L1*/
const Foo = () => {
  return (
    <>
      {" "}
      /*@pug:L2*/
      <div
        {...{
          "(click)": "play()",
        }}
        className={"div-class"}
      ></div>{" "}
      /*@pug:L3*/
      <div
        {...{
          "(click)": "play()",
        }}
        className={"div-class"}
      ></div>
    </>
  );
};
",
}
`;

exports[`Attributes Style attributes 1`] = `
{
  "line": 0,
  "nodes": [
    {
      "block": {
        "line": 2,
        "nodes": [
          {
            "attributeBlocks": [],
            "attrs": [
              {
                "column": 7,
                "line": 2,
                "mustEscape": true,
                "name": "style",
                "val": "{color: 'red', background: 'green'}",
              },
            ],
            "block": {
              "line": 2,
              "nodes": [],
              "type": "Block",
            },
            "column": 5,
            "isInline": true,
            "line": 2,
            "name": "a",
            "selfClosing": false,
            "type": "Tag",
          },
        ],
        "type": "Block",
      },
      "buffer": false,
      "column": 1,
      "isInline": false,
      "line": 1,
      "mustEscape": false,
      "type": "Code",
      "val": "const Foo = () =>",
    },
  ],
  "type": "Block",
}
`;

exports[`Attributes Style attributes 2`] = `
{
  "errors": [],
  "result": "/*@pug:L1*/
const Foo = () => (
  /*@pug:L2*/ <a
    style={{
      color: "red",
      background: "green",
    }}
  ></a>
);
",
}
`;

exports[`Case Default usage 1`] = `
{
  "line": 0,
  "nodes": [
    {
      "block": {
        "line": 2,
        "nodes": [
          {
            "buffer": false,
            "column": 5,
            "isInline": false,
            "line": 2,
            "mustEscape": false,
            "type": "Code",
            "val": "var friends = 10",
          },
          {
            "block": {
              "line": 4,
              "nodes": [
                {
                  "block": {
                    "line": 5,
                    "nodes": [
                      {
                        "attributeBlocks": [],
                        "attrs": [],
                        "block": {
                          "line": 5,
                          "nodes": [
                            {
                              "column": 15,
                              "line": 5,
                              "type": "Text",
                              "val": "you have no friends",
                            },
                          ],
                          "type": "Block",
                        },
                        "column": 13,
                        "isInline": false,
                        "line": 5,
                        "name": "p",
                        "selfClosing": false,
                        "type": "Tag",
                      },
                    ],
                    "type": "Block",
                  },
                  "column": 9,
                  "debug": false,
                  "expr": "0",
                  "line": 4,
                  "type": "When",
                },
                {
                  "block": {
                    "line": 7,
                    "nodes": [
                      {
                        "attributeBlocks": [],
                        "attrs": [],
                        "block": {
                          "line": 7,
                          "nodes": [
                            {
                              "column": 15,
                              "line": 7,
                              "type": "Text",
                              "val": "you have a friend",
                            },
                          ],
                          "type": "Block",
                        },
                        "column": 13,
                        "isInline": false,
                        "line": 7,
                        "name": "p",
                        "selfClosing": false,
                        "type": "Tag",
                      },
                    ],
                    "type": "Block",
                  },
                  "column": 9,
                  "debug": false,
                  "expr": "1",
                  "line": 6,
                  "type": "When",
                },
                {
                  "block": {
                    "line": 9,
                    "nodes": [
                      {
                        "attributeBlocks": [],
                        "attrs": [],
                        "block": {
                          "line": 9,
                          "nodes": [
                            {
                              "column": 15,
                              "line": 9,
                              "type": "Text",
                              "val": "you have ",
                            },
                            {
                              "buffer": true,
                              "column": 24,
                              "isInline": true,
                              "line": 9,
                              "mustEscape": true,
                              "type": "Code",
                              "val": "friends",
                            },
                            {
                              "column": 34,
                              "line": 9,
                              "type": "Text",
                              "val": " friends",
                            },
                          ],
                          "type": "Block",
                        },
                        "column": 13,
                        "isInline": false,
                        "line": 9,
                        "name": "p",
                        "selfClosing": false,
                        "type": "Tag",
                      },
                    ],
                    "type": "Block",
                  },
                  "column": 9,
                  "debug": false,
                  "expr": "default",
                  "line": 8,
                  "type": "When",
                },
              ],
              "type": "Block",
            },
            "column": 5,
            "expr": "friends",
            "line": 3,
            "type": "Case",
          },
        ],
        "type": "Block",
      },
      "buffer": false,
      "column": 1,
      "isInline": false,
      "line": 1,
      "mustEscape": false,
      "type": "Code",
      "val": "const Foo = () => ",
    },
  ],
  "type": "Block",
}
`;

exports[`Case Default usage 2`] = `
{
  "errors": [],
  "result": "/*@pug:L1*/
const Foo = () => {
  /*@pug:L2*/
  var friends = 10;
  return (
    <>
      {
        /*@pug:L3*/ (() => {
          switch (friends) {
            case 0:
              return /*@pug:L5*/ <p>you have no friends</p>;
            case 1:
              return /*@pug:L7*/ <p>you have a friend</p>;
            default:
              return /*@pug:L9*/ <p>you have {/*@pug:L9*/ friends} friends</p>;
          }
        })()
      }
    </>
  );
};
",
}
`;

exports[`Conditional Default usage 1`] = `
{
  "line": 0,
  "nodes": [
    {
      "block": {
        "line": 2,
        "nodes": [
          {
            "buffer": false,
            "column": 5,
            "isInline": false,
            "line": 2,
            "mustEscape": false,
            "type": "Code",
            "val": "var user = { description: 'foo bar baz' }",
          },
          {
            "buffer": false,
            "column": 5,
            "isInline": false,
            "line": 3,
            "mustEscape": false,
            "type": "Code",
            "val": "var authorised = false",
          },
          {
            "attributeBlocks": [],
            "attrs": [
              {
                "column": 5,
                "line": 4,
                "mustEscape": false,
                "name": "id",
                "val": "'user'",
              },
            ],
            "block": {
              "line": 4,
              "nodes": [
                {
                  "alternate": {
                    "alternate": {
                      "line": 14,
                      "nodes": [
                        {
                          "attributeBlocks": [],
                          "attrs": [
                            {
                              "column": 15,
                              "line": 14,
                              "mustEscape": false,
                              "name": "class",
                              "val": "'red'",
                            },
                          ],
                          "block": {
                            "line": 14,
                            "nodes": [
                              {
                                "column": 20,
                                "line": 14,
                                "type": "Text",
                                "val": "Description",
                              },
                            ],
                            "type": "Block",
                          },
                          "column": 13,
                          "isInline": false,
                          "line": 14,
                          "name": "h2",
                          "selfClosing": false,
                          "type": "Tag",
                        },
                        {
                          "attributeBlocks": [],
                          "attrs": [
                            {
                              "column": 14,
                              "line": 15,
                              "mustEscape": false,
                              "name": "class",
                              "val": "'description'",
                            },
                          ],
                          "block": {
                            "line": 15,
                            "nodes": [
                              {
                                "column": 27,
                                "line": 15,
                                "type": "Text",
                                "val": "User has no description",
                              },
                            ],
                            "type": "Block",
                          },
                          "column": 13,
                          "isInline": false,
                          "line": 15,
                          "name": "p",
                          "selfClosing": false,
                          "type": "Tag",
                        },
                      ],
                      "type": "Block",
                    },
                    "column": 9,
                    "consequent": {
                      "line": 9,
                      "nodes": [
                        {
                          "attributeBlocks": [],
                          "attrs": [
                            {
                              "column": 15,
                              "line": 9,
                              "mustEscape": false,
                              "name": "class",
                              "val": "'blue'",
                            },
                          ],
                          "block": {
                            "line": 9,
                            "nodes": [
                              {
                                "column": 21,
                                "line": 9,
                                "type": "Text",
                                "val": "Description",
                              },
                            ],
                            "type": "Block",
                          },
                          "column": 13,
                          "isInline": false,
                          "line": 9,
                          "name": "h2",
                          "selfClosing": false,
                          "type": "Tag",
                        },
                        {
                          "attributeBlocks": [],
                          "attrs": [
                            {
                              "column": 14,
                              "line": 10,
                              "mustEscape": false,
                              "name": "class",
                              "val": "'description'",
                            },
                          ],
                          "block": {
                            "line": 10,
                            "nodes": [
                              {
                                "column": 17,
                                "line": 11,
                                "type": "Text",
                                "val": "User has no descriptio",
                              },
                              {
                                "column": 1,
                                "line": 12,
                                "type": "Text",
                                "val": "
",
                              },
                              {
                                "column": 17,
                                "line": 12,
                                "type": "Text",
                                "val": "why not add one...",
                              },
                            ],
                            "type": "Block",
                          },
                          "column": 13,
                          "isInline": false,
                          "line": 10,
                          "name": "p",
                          "selfClosing": false,
                          "textOnly": true,
                          "type": "Tag",
                        },
                      ],
                      "type": "Block",
                    },
                    "line": 8,
                    "test": "authorised",
                    "type": "Conditional",
                  },
                  "column": 9,
                  "consequent": {
                    "line": 6,
                    "nodes": [
                      {
                        "attributeBlocks": [],
                        "attrs": [
                          {
                            "column": 15,
                            "line": 6,
                            "mustEscape": false,
                            "name": "class",
                            "val": "'green'",
                          },
                        ],
                        "block": {
                          "line": 6,
                          "nodes": [
                            {
                              "column": 22,
                              "line": 6,
                              "type": "Text",
                              "val": "Description",
                            },
                          ],
                          "type": "Block",
                        },
                        "column": 13,
                        "isInline": false,
                        "line": 6,
                        "name": "h2",
                        "selfClosing": false,
                        "type": "Tag",
                      },
                      {
                        "attributeBlocks": [],
                        "attrs": [
                          {
                            "column": 14,
                            "line": 7,
                            "mustEscape": false,
                            "name": "class",
                            "val": "'description'",
                          },
                        ],
                        "block": {
                          "line": 7,
                          "nodes": [
                            {
                              "buffer": true,
                              "column": 26,
                              "isInline": true,
                              "line": 7,
                              "mustEscape": true,
                              "type": "Code",
                              "val": "user.description",
                            },
                          ],
                          "type": "Block",
                        },
                        "column": 13,
                        "isInline": false,
                        "line": 7,
                        "name": "p",
                        "selfClosing": false,
                        "type": "Tag",
                      },
                    ],
                    "type": "Block",
                  },
                  "line": 5,
                  "test": "user.description",
                  "type": "Conditional",
                },
              ],
              "type": "Block",
            },
            "column": 5,
            "isInline": false,
            "line": 4,
            "name": "div",
            "selfClosing": false,
            "type": "Tag",
          },
        ],
        "type": "Block",
      },
      "buffer": false,
      "column": 1,
      "isInline": false,
      "line": 1,
      "mustEscape": false,
      "type": "Code",
      "val": "const Foo = () =>",
    },
  ],
  "type": "Block",
}
`;

exports[`Conditional Default usage 2`] = `
{
  "errors": [],
  "result": "/*@pug:L1*/
const Foo = () => {
  /*@pug:L2*/
  var user = {
    description: "foo bar baz",
  };
  /*@pug:L3*/
  var authorised = false;
  return (
    /*@pug:L4*/ <div id={"user"}>
      {
        /*@pug:L5*/ user.description ? (
          <>
            {" "}
            /*@pug:L6*/<h2 className={"green"}>Description</h2> /*@pug:L7*/
            <p className={"description"}>{/*@pug:L7*/ user.description}</p>
          </>
        ) : (
          <>
            {
              /*@pug:L8*/ authorised ? (
                <>
                  {" "}
                  /*@pug:L9*/<h2 className={"blue"}>Description</h2>{" "}
                  /*@pug:L10*/
                  <p className={"description"}>
                    User has no descriptio why not add one...
                  </p>
                </>
              ) : (
                <>
                  {" "}
                  /*@pug:L14*/<h2 className={"red"}>Description</h2>{" "}
                  /*@pug:L15*/
                  <p className={"description"}>User has no description</p>
                </>
              )
            }
          </>
        )
      }
    </div>
  );
};
",
}
`;

exports[`Each Iteration over object 1`] = `
{
  "line": 0,
  "nodes": [
    {
      "block": {
        "line": 2,
        "nodes": [
          {
            "attributeBlocks": [],
            "attrs": [],
            "block": {
              "line": 2,
              "nodes": [
                {
                  "block": {
                    "line": 4,
                    "nodes": [
                      {
                        "attributeBlocks": [],
                        "attrs": [],
                        "block": {
                          "line": 4,
                          "nodes": [
                            {
                              "buffer": true,
                              "column": 15,
                              "isInline": true,
                              "line": 4,
                              "mustEscape": true,
                              "type": "Code",
                              "val": "index + ': ' + val",
                            },
                          ],
                          "type": "Block",
                        },
                        "column": 13,
                        "isInline": false,
                        "line": 4,
                        "name": "li",
                        "selfClosing": false,
                        "type": "Tag",
                      },
                    ],
                    "type": "Block",
                  },
                  "column": 9,
                  "key": "index",
                  "line": 3,
                  "obj": "{1:'one',2:'two',3:'three'}",
                  "type": "Each",
                  "val": "val",
                },
              ],
              "type": "Block",
            },
            "column": 5,
            "isInline": false,
            "line": 2,
            "name": "ul",
            "selfClosing": false,
            "type": "Tag",
          },
        ],
        "type": "Block",
      },
      "buffer": false,
      "column": 1,
      "isInline": false,
      "line": 1,
      "mustEscape": false,
      "type": "Code",
      "val": "const Foo = () =>",
    },
  ],
  "type": "Block",
}
`;

exports[`Each Iteration over object 2`] = `
{
  "errors": [],
  "result": "/*@pug:L1*/
const Foo = () => (
  /*@pug:L2*/ <ul>
    {
      /*@pug:L3*/ map(
        {
          1: "one",
          2: "two",
          3: "three",
        },
        (val, index) => (
          /*@pug:L4*/ <li>{/*@pug:L4*/ index + ": " + val}</li>
        )
      )
    }
  </ul>
);
",
}
`;

exports[`Each Iteration with member 1`] = `
{
  "line": 0,
  "nodes": [
    {
      "block": {
        "line": 2,
        "nodes": [
          {
            "attributeBlocks": [],
            "attrs": [],
            "block": {
              "line": 2,
              "nodes": [
                {
                  "block": {
                    "line": 4,
                    "nodes": [
                      {
                        "attributeBlocks": [],
                        "attrs": [],
                        "block": {
                          "line": 4,
                          "nodes": [
                            {
                              "buffer": true,
                              "column": 15,
                              "isInline": true,
                              "line": 4,
                              "mustEscape": true,
                              "type": "Code",
                              "val": "val",
                            },
                          ],
                          "type": "Block",
                        },
                        "column": 13,
                        "isInline": false,
                        "line": 4,
                        "name": "li",
                        "selfClosing": false,
                        "type": "Tag",
                      },
                    ],
                    "type": "Block",
                  },
                  "column": 9,
                  "key": null,
                  "line": 3,
                  "obj": "[1, 2, 3, 4, 5]",
                  "type": "Each",
                  "val": "val",
                },
              ],
              "type": "Block",
            },
            "column": 5,
            "isInline": false,
            "line": 2,
            "name": "ul",
            "selfClosing": false,
            "type": "Tag",
          },
        ],
        "type": "Block",
      },
      "buffer": false,
      "column": 1,
      "isInline": false,
      "line": 1,
      "mustEscape": false,
      "type": "Code",
      "val": "const Foo = () =>",
    },
  ],
  "type": "Block",
}
`;

exports[`Each Iteration with member 2`] = `
{
  "errors": [],
  "result": "/*@pug:L1*/
const Foo = () => (
  /*@pug:L2*/ <ul>
    {
      /*@pug:L3*/ map([1, 2, 3, 4, 5], (val) => (
        /*@pug:L4*/ <li>{/*@pug:L4*/ val}</li>
      ))
    }
  </ul>
);
",
}
`;

exports[`Each Iteration with member and index 1`] = `
{
  "line": 0,
  "nodes": [
    {
      "block": {
        "line": 2,
        "nodes": [
          {
            "attributeBlocks": [],
            "attrs": [],
            "block": {
              "line": 2,
              "nodes": [
                {
                  "block": {
                    "line": 4,
                    "nodes": [
                      {
                        "attributeBlocks": [],
                        "attrs": [],
                        "block": {
                          "line": 4,
                          "nodes": [
                            {
                              "buffer": true,
                              "column": 15,
                              "isInline": true,
                              "line": 4,
                              "mustEscape": true,
                              "type": "Code",
                              "val": "index + ': ' + val",
                            },
                          ],
                          "type": "Block",
                        },
                        "column": 13,
                        "isInline": false,
                        "line": 4,
                        "name": "li",
                        "selfClosing": false,
                        "type": "Tag",
                      },
                    ],
                    "type": "Block",
                  },
                  "column": 9,
                  "key": "index",
                  "line": 3,
                  "obj": "['zero', 'one', 'two']",
                  "type": "Each",
                  "val": "val",
                },
              ],
              "type": "Block",
            },
            "column": 5,
            "isInline": false,
            "line": 2,
            "name": "ul",
            "selfClosing": false,
            "type": "Tag",
          },
        ],
        "type": "Block",
      },
      "buffer": false,
      "column": 1,
      "isInline": false,
      "line": 1,
      "mustEscape": false,
      "type": "Code",
      "val": "const Foo = () =>",
    },
  ],
  "type": "Block",
}
`;

exports[`Each Iteration with member and index 2`] = `
{
  "errors": [],
  "result": "/*@pug:L1*/
const Foo = () => (
  /*@pug:L2*/ <ul>
    {
      /*@pug:L3*/ map(["zero", "one", "two"], (val, index) => (
        /*@pug:L4*/ <li>{/*@pug:L4*/ index + ": " + val}</li>
      ))
    }
  </ul>
);
",
}
`;
